In [49]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px


In [50]:
data = pd.read_csv('open_meteo.csv')
data

Unnamed: 0        time  weathercode  temperature_2m_max  \
0               0  2010-01-01            0                26.0   
1               1  2010-01-02            2                25.3   
2               2  2010-01-03            0                25.0   
3               3  2010-01-04            0                25.6   
4               4  2010-01-05            0                25.4   
...           ...         ...          ...                 ...   
34757       34757  2023-08-02           51                30.2   
34758       34758  2023-08-03           51                30.4   
34759       34759  2023-08-04           51                30.5   
34760       34760  2023-08-05           51                30.7   
34761       34761  2023-08-06           51                30.7   

       temperature_2m_min  temperature_2m_mean  apparent_temperature_max  \
0                    14.8                 20.9                      24.0   
1                    15.6                 20.5                      21.8   
2                    14.3                 19.9                      22.8   
3                    15.0                 20.4                      23.9   
4                    16.3                 20.7                      24.5   
...                   ...                  ...                       ...   
34757                25.5                 27.4                      31.5   
34758                25.4                 27.3                      31.5   
34759                25.5                 27.5                      32.8   
34760                25.6                 27.5                      32.5   
34761                25.4                 27.3                      31.7   

       apparent_temperature_min  apparent_temperature_mean           sunrise  \
0                          13.4                       18.6  2010-01-01T02:16   
1                          13.7                       18.1  2010-01-02T02:17   
2                          14.2                       18.9  2010-01-03T02:17   
3                          13.7                       19.4  2010-01-04T02:17   
4                          16.8                       20.7  2010-01-05T02:17   
...                         ...                        ...               ...   
34757                      27.8                       29.3  2023-08-02T00:52   
34758                      28.2                       29.7  2023-08-03T00:52   
34759                      29.3                       30.6  2023-08-04T00:52   
34760                      28.9                       30.4  2023-08-05T00:53   
34761                      27.0                       28.9  2023-08-06T00:53   

       ... precipitation_hours  windspeed_10m_max  windgusts_10m_max  \
0      ...                 0.0               18.3               25.6   
1      ...                 0.0               24.6               35.3   
2      ...                 0.0               16.6               28.1   
3      ...                 0.0               14.0               22.7   
4      ...                 0.0               19.3               29.2   
...    ...                 ...                ...                ...   
34757  ...                 6.0               37.7               58.3   
34758  ...                10.0               36.4               56.2   
34759  ...                 6.0               31.8               49.3   
34760  ...                 6.0               30.7               48.6   
34761  ...                 4.0               38.8               60.5   

       winddirection_10m_dominant  et0_fao_evapotranspiration   latitude  \
0                              54                        4.38  24.900002   
1                              64                        3.84  24.900002   
2                             312                        3.69  24.900002   
3                             353                        3.63  24.900002   
4                             270                        3.55  24.900002   
...                

In [51]:
data.describe()

Unnamed: 0   weathercode  temperature_2m_max  temperature_2m_min  \
count  34762.000000  34762.000000        34762.000000        34762.000000   
mean   17380.500000     24.229417           30.677150           23.205748   
std    10035.069365     27.551242            3.236343            3.615309   
min        0.000000      0.000000           17.700000            6.800000   
25%     8690.250000      1.000000           28.300000           21.700000   
50%    17380.500000      2.000000           30.500000           24.200000   
75%    26070.750000     53.000000           32.800000           25.500000   
max    34761.000000     65.000000           44.300000           32.500000   

       temperature_2m_mean  apparent_temperature_max  \
count         34762.000000              34762.000000   
mean             26.674475                 33.920612   
std               2.733634                  4.309410   
min              13.200000                 11.800000   
25%              25.500000                 31.800000   
50%              26.800000                 34.400000   
75%              28.300000                 36.800000   
max              37.100000                 47.100000   

       apparent_temperature_min  apparent_temperature_mean  \
count              34762.000000               34762.000000   
mean                  26.026771                  29.564956   
std                    5.710073                   4.509885   
min                    1.400000                   7.600000   
25%                   23.500000                  28.200000   
50%                   27.900000                  30.500000   
75%                   29.800000                  32.400000   
max                   37.400000                  41.100000   

       shortwave_radiation_sum  precipitation_sum      rain_sum  snowfall_sum  \
count             34762.000000       34762.000000  34762.000000       34762.0   
mean                 19.105281           4.496036      4.496036           0.0   
std                   5.091534          12.391169     12.391169           0.0   
min                   0.720000           0.000000      0.000000           0.0   
25%                  16.670000           0.000000      0.000000           0.0   
50%                  19.570000           0.000000      0.000000           0.0   
75%                  22.610000           2.800000      2.800000           0.0   
max                  29.320000         307.900000    307.900000           0.0   

       precipitation_hours  windspeed_10m_max  windgusts_10m_max  \
count         34762.000000       34762.000000       34762.000000   
mean              5.321788          18.603832          34.039563   
std               8.286665           6.353607           9.502845   
min               0.000000           4.000000          13.700000   
25%               0.000000          14.000000          27.000000   
50%               0.000000          17.300000          32.800000   
75%               9.000000          22.000000          39.200000   
max              24.000000          78.700000         123.500000   

       winddirection_10m_dominant  et0_fao_evapotranspiration      latitude  \
count                34762.000000                34762.000000  34762.000000   
mean                   224.866089                    4.568428     17.900002   
std                     98.786400                    1.553584      5.478609   
min                      0.000000                    0.700000      8.500000   
25%                    193.000000                    3.620000     12.900002   
50%                    251.000000                    4.520000     19.099998   
75%                    297.000000                    5.460000     23.300003   
max                    360.000000                   13.280000     24.900002   

          longitude     elevation  
count  34762.000000  34762.000000  
mean      72.585718     26.000000  
std        3.055919     32.789843  
min       67.000000      6.000000  
25%       69.700

In [52]:
data = data[['time', 'city', 'temperature_2m_min', 'temperature_2m_max', 'temperature_2m_mean', 'precipitation_sum', 'precipitation_hours', 'et0_fao_evapotranspiration']]
data

time     city  temperature_2m_min  temperature_2m_max  \
0      2010-01-01  Karachi                14.8                26.0   
1      2010-01-02  Karachi                15.6                25.3   
2      2010-01-03  Karachi                14.3                25.0   
3      2010-01-04  Karachi                15.0                25.6   
4      2010-01-05  Karachi                16.3                25.4   
...           ...      ...                 ...                 ...   
34757  2023-08-02     Bhuj                25.5                30.2   
34758  2023-08-03     Bhuj                25.4                30.4   
34759  2023-08-04     Bhuj                25.5                30.5   
34760  2023-08-05     Bhuj                25.6                30.7   
34761  2023-08-06     Bhuj                25.4                30.7   

       temperature_2m_mean  precipitation_sum  precipitation_hours  \
0                     20.9                0.0                  0.0   
1                     20.5                0.0                  0.0   
2                     19.9                0.0                  0.0   
3                     20.4                0.0                  0.0   
4                     20.7                0.0                  0.0   
...                    ...                ...                  ...   
34757                 27.4                0.6                  6.0   
34758                 27.3                1.0                 10.0   
34759                 27.5                0.8                  6.0   
34760                 27.5                0.6                  6.0   
34761                 27.3                0.4                  4.0   

       et0_fao_evapotranspiration  
0                            4.38  
1                            3.84  
2                            3.69  
3                            3.63  
4                            3.55  
...                           ...  
34757                        4.07  
34758                        3.92  
34759                        3.75  
34760                        4.10  
34761                        4.31  

[34762 rows x 8 columns]

In [53]:
data['time'] = pd.to_datetime(data['time'], format='%Y/%m/%d')


C:\Users\ASUS\anaconda3\lib\site-packages\pandas\core\frame.py:3607: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [54]:
start_date = '5-1-2022'
end_date = '12-1-2022'
df = data[(data['time'] >= start_date) & (data['time'] <= end_date)]

In [55]:
df

time     city  temperature_2m_min  temperature_2m_max  \
4503  2022-05-01  Karachi                28.0                36.2   
4504  2022-05-02  Karachi                26.6                33.1   
4505  2022-05-03  Karachi                26.1                33.5   
4506  2022-05-04  Karachi                25.5                32.9   
4507  2022-05-05  Karachi                25.0                33.0   
...          ...      ...                 ...                 ...   
34509 2022-11-27     Bhuj                15.4                29.8   
34510 2022-11-28     Bhuj                14.7                29.8   
34511 2022-11-29     Bhuj                14.1                29.4   
34512 2022-11-30     Bhuj                15.0                30.0   
34513 2022-12-01     Bhuj                14.9                29.0   

       temperature_2m_mean  precipitation_sum  precipitation_hours  \
4503                  31.8                0.0                  0.0   
4504                  29.5                0.0                  0.0   
4505                  29.4                0.0                  0.0   
4506                  28.7                0.0                  0.0   
4507                  28.3                0.0                  0.0   
...                    ...                ...                  ...   
34509                 22.8                0.0                  0.0   
34510                 22.3                0.0                  0.0   
34511                 21.8                0.0                  0.0   
34512                 22.6                0.0                  0.0   
34513                 22.2                0.0                  0.0   

       et0_fao_evapotranspiration  
4503                         9.70  
4504                         7.83  
4505                         6.71  
4506                         6.50  
4507                         6.32  
...                           ...  
34509                        4.05  
34510                        4.48  
34511                        5.08  
34512                        5.41  
34513                        4.81  

[1505 rows x 8 columns]

### Which city has low temperature and high temperature? 

In [56]:
low_temp = df['temperature_2m_min'].min()
df[df['temperature_2m_min'] == low_temp]

time  city  temperature_2m_min  temperature_2m_max  \
34511 2022-11-29  Bhuj                14.1                29.4   

       temperature_2m_mean  precipitation_sum  precipitation_hours  \
34511                 21.8                0.0                  0.0   

       et0_fao_evapotranspiration  
34511                        5.08

We can see that Bhuj has low temperature than the other cities.
The city was not too hot at 2022 November

In [57]:
high_temp = df['temperature_2m_max'].max()
df[df['temperature_2m_max'] == high_temp] 

time  city  temperature_2m_min  temperature_2m_max  \
34310 2022-05-12  Bhuj                25.2                42.2   
34353 2022-06-24  Bhuj                27.8                42.2   

       temperature_2m_mean  precipitation_sum  precipitation_hours  \
34310                 32.9                0.0                  0.0   
34353                 34.6                0.0                  0.0   

       et0_fao_evapotranspiration  
34310                       10.16  
34353                        8.29

From the above, The maximum temperature recorded in Bhuj was 42.2°C. The months of May and June are indeed known for higher temperatures and heatwaves in Bhuj.

### Minimum Temperature and Maximum Temperature Plot for 2022

In [58]:
fig = px.line(df, x="time", y="temperature_2m_min", title='Minimum Temperature for each day',color='city')
fig2 = px.line(df, x="time", y="temperature_2m_max", title='Maximum Temperature for each day',color='city')
fig3 = px.line(df, x="time", y="temperature_2m_mean", title='Average Temperature for each day',color='city')

fig.show()
fig2.show()
fig3.show()

From the above plot, we can see that all of the cities temperature. For instance, in 2020, in July, the minimum temperature is 28.9 degrees Celsius and the high temperature is 30 degrees Celsius in May. From these plots, we can conclude that the high temperature occurs around May, June, and October.

### Minimum Temperature and Maximum Temperature Plot for 2021

In [59]:
start_date = '5-1-2021'
end_date = '12-1-2021'
df2 = data[(data['time'] >= start_date) & (data['time'] <= end_date)]

In [60]:
fig = px.line(df2, x="time", y="temperature_2m_min", title='Minimum Temperature for each day',color='city')
fig2 = px.line(df2, x="time", y="temperature_2m_max", title='Maximum Temperature for each day',color='city')
fig.show()
fig2.show()

From the above graphs, we can clearly see that the high temperature in December 2021 was 30.1 degrees Celsius. From the previous graph, the high temperature in December 2022 was 28.7 degrees Celsius. That means last year's temperature was not as hot as 2021.

In [61]:
df

time     city  temperature_2m_min  temperature_2m_max  \
4503  2022-05-01  Karachi                28.0                36.2   
4504  2022-05-02  Karachi                26.6                33.1   
4505  2022-05-03  Karachi                26.1                33.5   
4506  2022-05-04  Karachi                25.5                32.9   
4507  2022-05-05  Karachi                25.0                33.0   
...          ...      ...                 ...                 ...   
34509 2022-11-27     Bhuj                15.4                29.8   
34510 2022-11-28     Bhuj                14.7                29.8   
34511 2022-11-29     Bhuj                14.1                29.4   
34512 2022-11-30     Bhuj                15.0                30.0   
34513 2022-12-01     Bhuj                14.9                29.0   

       temperature_2m_mean  precipitation_sum  precipitation_hours  \
4503                  31.8                0.0                  0.0   
4504                  29.5                0.0                  0.0   
4505                  29.4                0.0                  0.0   
4506                  28.7                0.0                  0.0   
4507                  28.3                0.0                  0.0   
...                    ...                ...                  ...   
34509                 22.8                0.0                  0.0   
34510                 22.3                0.0                  0.0   
34511                 21.8                0.0                  0.0   
34512                 22.6                0.0                  0.0   
34513                 22.2                0.0                  0.0   

       et0_fao_evapotranspiration  
4503                         9.70  
4504                         7.83  
4505                         6.71  
4506                         6.50  
4507                         6.32  
...                           ...  
34509                        4.05  
34510                        4.48  
34511                        5.08  
34512                        5.41  
34513                        4.81  

[1505 rows x 8 columns]

### Most heat waves period

In [62]:
temperature_threshold = 30 
heatwave_periods = df[df['temperature_2m_max'] > temperature_threshold]
heatwave_periods

time     city  temperature_2m_min  temperature_2m_max  \
4503  2022-05-01  Karachi                28.0                36.2   
4504  2022-05-02  Karachi                26.6                33.1   
4505  2022-05-03  Karachi                26.1                33.5   
4506  2022-05-04  Karachi                25.5                32.9   
4507  2022-05-05  Karachi                25.0                33.0   
...          ...      ...                 ...                 ...   
34502 2022-11-20     Bhuj                17.1                30.6   
34503 2022-11-21     Bhuj                17.1                30.7   
34504 2022-11-22     Bhuj                16.5                30.5   
34505 2022-11-23     Bhuj                16.3                30.5   
34506 2022-11-24     Bhuj                16.2                30.1   

       temperature_2m_mean  precipitation_sum  precipitation_hours  \
4503                  31.8                0.0                  0.0   
4504                  29.5                0.0                  0.0   
4505                  29.4                0.0                  0.0   
4506                  28.7                0.0                  0.0   
4507                  28.3                0.0                  0.0   
...                    ...                ...                  ...   
34502                 24.4                0.0                  0.0   
34503                 24.2                0.0                  0.0   
34504                 23.9                0.0                  0.0   
34505                 23.4                0.0                  0.0   
34506                 22.9                0.0                  0.0   

       et0_fao_evapotranspiration  
4503                         9.70  
4504                         7.83  
4505                         6.71  
4506                         6.50  
4507                         6.32  
...                           ...  
34502                        4.84  
34503                        4.80  
34504                        4.63  
34505                        4.78  
34506                        4.82  

[691 rows x 8 columns]

In [63]:
px.line(heatwave_periods,x='time',y='temperature_2m_max',title='Heat waves periods',color='city')

Bhuj city causes heat waves other than those in other countries. Karachi doesn't have as many heatwaves as Bhuj. The high temperatures occur in May and June.

### Correlation for each city

In [64]:
df.groupby('city').temperature_2m_min.describe()

count       mean       std   min    25%   50%   75%   max
city                                                                 
Bhuj        215.0  23.886512  3.043402  14.1  23.15  24.9  25.8  28.6
Karachi     215.0  25.377209  2.519400  15.8  24.75  26.3  27.0  28.9
Mangalore   215.0  24.253953  1.036110  20.6  23.80  24.3  24.7  28.0
Mumbai      215.0  25.008837  1.965917  19.4  24.50  25.4  25.9  28.4
Old Goa     215.0  24.495349  1.755705  18.5  24.10  24.8  25.3  28.9
Surat       215.0  25.037674  2.410341  17.4  24.80  25.6  26.5  28.5
Trivandrum  215.0  23.800000  0.646414  22.0  23.50  23.8  24.1  26.3

Temperatures range from 18.5°C to 28.9°C, with an average around 24.5°C. The temperature distribution is somewhat narrow, with a standard deviation of 1.76°C.

In [65]:
df.groupby('city').temperature_2m_max.describe()

count       mean       std   min    25%   50%    75%   max
city                                                                  
Bhuj        215.0  33.297209  3.636303  25.6  30.45  32.2  35.85  42.2
Karachi     215.0  31.533953  2.143652  27.7  29.85  31.3  33.00  38.0
Mangalore   215.0  28.545581  1.706137  25.1  27.40  28.6  29.60  33.2
Mumbai      215.0  29.600465  2.053251  25.7  27.90  29.4  31.35  34.4
Old Goa     215.0  30.548837  2.598397  25.4  28.45  30.2  32.80  36.4
Surat       215.0  30.652558  2.211351  26.4  28.95  30.4  32.30  37.6
Trivandrum  215.0  27.556279  1.002168  24.6  27.05  27.6  28.10  30.5

The mean maximum temperature is about 31.5°C, with temperatures ranging from 27.7°C to 38.0°C. The distribution is relatively tight, with a standard deviation of 2.14°C.

In [76]:
city_corr = pd.pivot_table(df, values='temperature_2m_max', index=['time'], columns=['city'])
city_corr

city        Bhuj  Karachi  Mangalore  Mumbai   Old Goa  Surat  Trivandrum
time                                                                     
2022-05-01  40.9     36.2       31.5     32.9     33.8   34.0        29.8
2022-05-02  39.0     33.1       32.6     32.8     34.1   34.1        29.9
2022-05-03  36.0     33.5       33.2     33.1     34.8   33.6        30.5
2022-05-04  37.5     32.9       32.5     32.7     36.4   34.0        29.8
2022-05-05  38.3     33.0       31.2     33.0     34.6   34.4        29.6
...          ...      ...        ...      ...      ...    ...         ...
2022-11-27  29.8     27.7       29.2     30.3     31.3   29.8        27.4
2022-11-28  29.8     28.8       29.4     30.1     30.5   29.2        28.0
2022-11-29  29.4     29.6       29.9     29.6     30.9   30.2        28.0
2022-11-30  30.0     29.7       30.9     31.1     34.0   30.7        28.4
2022-12-01  29.0     28.7       30.7     30.8     33.9   30.1        28.4

[215 rows x 7 columns]

In [77]:
city_corr.corr()

city            Bhuj   Karachi  Mangalore   Mumbai    Old Goa     Surat  \
city                                                                      
Bhuj        1.000000  0.733117   0.257518  0.664624  0.376890  0.806565   
Karachi     0.733117  1.000000  -0.039880  0.301256  0.055800  0.398236   
Mangalore   0.257518 -0.039880   1.000000  0.647333  0.858273  0.525153   
Mumbai      0.664624  0.301256   0.647333  1.000000  0.767799  0.892983   
Old Goa     0.376890  0.055800   0.858273  0.767799  1.000000  0.636389   
Surat       0.806565  0.398236   0.525153  0.892983  0.636389  1.000000   
Trivandrum  0.269952  0.099011   0.466729  0.252060  0.293971  0.294217   

city        Trivandrum  
city                    
Bhuj          0.269952  
Karachi       0.099011  
Mangalore     0.466729  
Mumbai        0.252060  
Old Goa       0.293971  
Surat         0.294217  
Trivandrum    1.000000

The correlation between Bhuj and Karachi has high correlation based on maximum temperature. When the maximum temperature in Bhuj tends to increase, the maximum temperature in Karachi also tends to increase, and vice versa.